### PR TITLE
Implement AsyncContext class

### PR DESCRIPTION
**Pull request by romandev**
This class provides a wrapper for the following custom asynchronous
operation APIs.
  - napi_async_init()
  - napi_async_destroy()
  - napi_make_callback()

This PR is initiated from https://github.com/nodejs/node-addon-api/pull/140#issuecomment-330685533.

### CODE REVIEW

**Comment by mhdawson** [comment]
@romandev are you going to have a chance to get back to this?

**Comment by romandev** [comment]
@mhdawson, I'll update this patch in this week.

File: doc/async_context.md [hunk 1]
```diff
@@ -0,0 +1,149 @@
+# AsyncContext
+
+The [AsyncWorker](async_worker.md) class may not be appropriate for every scenario, because with
+those the async execution still happens on the main event loop. When using any
```
Comment (mhdawson) [inline doc/async_context.md]: I don't think this is quite right as the async execution happens in a thread from the existing thread pool. I'll have a closer look and add some I think describes the difference.

Comment (romandev) [reply doc/async_context.md]: I matched up with https://nodejs.org/api/n-api.html#n_api_custom_asynchronous_operations.

**Comment by romandev** [comment]
@mhdawson @NickNaso Could you pleaser review this?

**Comment by romandev** [comment]
@mhdawson @NickNaso Gentle ping

**Comment by NickNaso** [comment]
@mhdawson Could you review this PR?

**Comment by mhdawson** [comment]
@romandev sorry its taken so long to comment.

A couple of high level comments/questions

1) It might be better if the class was not used with new/delete, but instead stack allocated and then automatically was cleaned up when existing a {} scope.  I think that is the way that HandleScope works.

2) I'm not sure including passing a callback and including the MakeCallback methods makes sense.  I think it would be possible to make 2 callbacks within a since AsyncContext, and if you already have a FunctionReference its just as easy to call MakeCallback on that?

**Comment by romandev** [comment]
@NickNaso, @mhdawson Sorry for delay. I've been recently busy for other works but I'll take this work in this weekend. I know that this is release blocker. If you don't mind, can you wait for me? or if you urgent, please feel free to take this.

**Comment by mhdawson** [comment]
@romandev if you can get to it this weekend I think it's worth waiting until next week for the release.

**Comment by romandev** [comment]
@mhdawson, @NickNaso I addressed all your comments. Please take another looks?

File: doc/function_reference.md [hunk 2]
```diff
@@ -187,12 +191,16 @@ Calls a referenced JavaScript function from a native add-on after an asynchronou
...
-Napi::Value Napi::FunctionReference::MakeCallback(napi_value recv, const std::vector<napi_value>& args) const;
+Napi::Value Napi::FunctionReference::MakeCallback(napi_value recv, const std::vector<napi_value>& args, napi_async_context context = nullptr) const;
 ```
 
 - `[in] recv`: The `this` object passed to the referenced function when it's called.
 - `[in] args`: Vector of JavaScript values as `napi_value` representing the
 arguments of the referenced function.
+- `[in] context`: Context for the async operation that is invoking the callback.
+This should normally be a value previously obtained from [Napi::AsyncContext](async_context.md).
+However `nullptr` is also allowed, which indicates the currenc async context
```
Comment (NickNaso) [inline doc/function_reference.md]: currenc -> current

File: doc/function_reference.md [hunk 3]
```diff
@@ -203,13 +211,17 @@ Calls a referenced JavaScript function from a native add-on after an asynchronou
...
+Napi::Value Napi::FunctionReference::MakeCallback(napi_value recv, size_t argc, const napi_value* args, napi_async_context context = nullptr) const;
 ```
 
 - `[in] recv`: The `this` object passed to the referenced function when it's called.
 - `[in] argc`: The number of arguments passed to the referenced function.
 - `[in] args`: Array of JavaScript values as `napi_value` representing the
 arguments of the referenced function.
+- `[in] context`: Context for the async operation that is invoking the callback.
+This should normally be a value previously obtained from [Napi::AsyncContext](async_context.md).
+However `nullptr` is also allowed, which indicates the currenc async context
```
Comment (NickNaso) [inline doc/function_reference.md]: currenc -> current

**Comment by romandev** [comment]
FYI, I fixed typo as well :)

**Comment by NickNaso** [comment]
@mhdawson Could you take a look at this PR?

**Comment by mhdawson** [comment]
CI run to make sure its ok on 6.x https://ci.nodejs.org/view/x%20-%20Abi%20stable%20module%20API/job/node-test-node-addon-api/784/

**Comment by mhdawson** [comment]
Landed as dfcb93945f96d6e2a01fce0d862af446b70d4f73

**Pull request closed without merging**